### PR TITLE
Disabled messaging to blocked and deleted users - fixes #9247 and #9989

### LINKED ISSUE #9247: prevent typing a private message to a player who has blocked you / opted out of PMs
On the Messages screen (shown below), when the player that you're trying to send a message to has blocked you or has opted out of all private messages, then there should be no "send" button or new message field at the bottom of the screen. Instead, there should be a statement that you can't message that person (use the `notAuthorizedToSendMessageToThisUser` locale string, which already exists: "You can't send a message to this player because they have chosen to block messages.").

However, mods/staff should always see the send button and the text field, regardless of whether the user has opted out or blocked the mod, because a partial fix for https://github.com/HabitRPG/habitica/issues/8142 allows mods/staff to PM all users regardless of blocking. Moderators and staff can be recognised in the code by `anyone with contributor.admin` being set to `true`.

image

_edit:_

I've logged https://github.com/HabitRPG/habitica/issues/9989 for a similar issue: "prevent typing a private message to a player who has deleted their account"
It could be either worked on at the same time as this issue, or after this issue is fixed, https://github.com/HabitRPG/habitica/issues/9989 could be fixed by reusing most of the new code.

### Note
Some work towards fixing this issue was already made in https://github.com/HabitRPG/habitica/pull/10878 and can be used as a starting point.

### LINKED ISSUE #9989: prevent typing a private message to a player who has deleted their account
This is similar to https://github.com/HabitRPG/habitica/issues/9247 ("prevent typing a private message to a player who has blocked you / opted out of PMs") and the same code might be usable for both.

On the Messages screen, when the player that you're trying to send a message to has deleted their account, then there should be no "send" button or new message field at the bottom of the screen. Instead, there should be a statement that you can't message that person because they have deleted their Habitica account (you'll need to create a new string for this and it should be placed in
the `groups.json` locales file, directly underneat the `notAuthorizedToSendMessageToThisUser` string to keep them both together).

Unlike https://github.com/HabitRPG/habitica/issues/9247 , there should be no special treatment for mods.

----

For a description of why this change is needed, here's a bug reportt from CakeKnight (d8fc8d8b-a10f-416a-b089-587db2af9e4d):

_"In User icon > Messages I tried to reply to an old conversation with another user and got an error saying the user was not found. The user's avatar is not loading, there are a few errors in the console, also when trying to click on the user's name or avatar to see their profile, but from what I see in the error message I'm pretty sure the user actually deleted their account._ [I confirmed that they had] _Also, it looks it's possible to send more messages to the deleted user, but after a refresh they're gone. It looks like a simple network error so for the average Habitican it will be difficult to understand what's going on, on the other hand a "this user deleted their account" would be pretty sad..."_

I agree that the deleted message would be sad but I think it's worth explaining to the user exactly why they can't message the user any more, or see their profile or avatar.

### Note
Some work towards fixing this issue was already made in https://github.com/HabitRPG/habitica/pull/10878 and can be used as a starting point.

### PR DESCRIPTION
**Pull request by nikkigabbard**
[//]: # (Note: See http://habitica.wikia.com/wiki/Using_Your_Local_Install_to_Modify_Habitica%27s_Website_and_API for more info)

[//]: # (Put Issue # here, if applicable. This will automatically close the issue if your PR is merged in)
Fixes #9247
Fixes #9989

### Changes
[//]: # (Describe the changes that were made in detail here. Include pictures if necessary)
image
image

Fixed the issue where users could message other users who had blocked them or message users who had deleted their account. This is done by disabling the text area as well as the send button and displaying an error card in the top right hand corner of the screen. 

[//]: # (Put User ID in here - found on the Habitica website at User Icon > Settings > API)

----
UUID: dd0379e7-4ca5-4ab1-a639-946dcf5b2ca9

### CODE REVIEW

**Comment by nikkigabbard** [comment]
issues #9247 and #9989

**Comment by Alys** [comment]
@nikkigabbard Thanks for submitting this! I've edited the top post to replace "Fixes put_#_and_issue_numer_here" with the issue numbers (also in the title). Can you please edit it to add your User ID (the 36-character string from the Settings > API page in the website) and to write a description of the change? Please also include screenshots of any visible changes to the website. See [Submitting your Change for Review in a Pull Request](https://habitica.wikia.com/wiki/Using_Your_Local_Install_to_Modify_Habitica's_Website_and_API#Submitting_your_Change_for_Review_in_a_Pull_Request) (step 4) for details about what's helpful.

File: .gitignore [hunk 1]
```diff
@@ -13,6 +13,7 @@ apidoc_build
 *.swp
 .idea*
 config.json
+package-lock.json
```
Comment (paglias) [inline .gitignore:16]: can you remove this?

File: test/api/unit/libs/cron.test.js [hunk 2]
```diff
@@ -646,7 +646,7 @@ describe('cron', () => {
       expect(user.purchased.plan.customerId).to.not.exist;
     });
 
-    xit('does nothing to plan stats when we are after the last day of the cancelled month', () => {
+    it('does nothing to plan stats when we are after the last day of the cancelled month', () => {
```
Comment (paglias) [inline test/api/unit/libs/cron.test.js:649]: can you revert the changes to the excluded tests? Thanks!

**Review by paglias — CHANGES_REQUESTED** [review]

**Comment by Tressley** [comment]
@nikkigabbard -- Thank you for your work on this! We actually have updated designs for the Inbox and how better to handle these use cases. Rather than contain the error message in a snack notification, we prefer it's displayed clearly in the Inbox. I've included some screenshots for a quick reference.

image
image

If you'd like to make these updates, I can invite you to our Zeplin project (where we house designs for hand-off). I'll just need you to share your email.

**Comment by paglias** [comment]
hi @nikkigabbard ! do you have any update on the status of this pr? thanks!

**Comment by paglias** [comment]
Hi @nikkigabbard I'm going to close this one but if you want to work on it just leave a comment

**Pull request closed without merging**